[menu zoom button] large buttons on raspberry pi which has touch screen

CazYokoyama opened the pull request:
As the comment below for HasTouchScreen() in Asset.hpp, we need bigger icons for touch screen. 

a touch screen may require bigger areas.

defined(RASPBERRY_PI) does not guarantee whether touch screen is attached. sysfs has an entry for the touch screen as below.
$ cat /sys/class/input/event?/device/name 
.....
1-0014 Goodix Capacitive TouchScreen
There are 2 ways to determine whether a touch screen is attached, C code and
`grep -i TouchScreen /sys/class/input/event?/device/name`
On either cases, we need fairly large amount of code. Let me know your preference, this simple code, C code or run grep.

CazYokoyama commented:
Please hold. Afterward test shows the code does not work. Will update.

CazYokoyama commented:
Can I assume the person who wants menu and zoom buttons has touch screen? Those buttons are clicked by mouse and touch screen. Larger icons cover large area of the map, but who care? You can disable them by configuration.

lordfolken commented:
We should have a  real touchscreen detection for Raspberry PI. A lot of the code depends on this and is imo good philosophy. 

I for example don't have a touch screen with mine, because i don't want fingermarks all over it and also rather have the increased contrast than a filter in front of the screen. Also personally i find button presses i can do without looking at the screen, far superior in a cockpit setting. (But i might be alone with that opinion.) 

About activating this by default:

These buttons are now the 4th way of interacting with xcsoar: (for example map zoom)
* Nav->Display-> Zoom in/out
* Gesture (up / down)
* Volume Rocker on a Phone / Keyboard / Joystick on others
* OnScreen Buttons

None of these (except the on-screen buttons) are self-explanatory and intuitive or build on familiarity of typical smartphone apps. For exmaple the first thing people try when i give xcsoar to them is pinch for zoom. 

Its also incredibly inconsistent. On the PAN page you can´t use gestures to zoom, and then there is onscreen buttons (others than the mainpage). 

I think we need a clear design guideline when it comes to interaction with xcsoar.

CazYokoyama committed:
[menu zoom button] add checking of sysfs for touchscreen in raspberry pi.
When xcsoar is invoked without X server, i.e. no desktop GUI, text console
event_queue->HasTouchScreen() reports false even it is attached.
Need to look for another way.

"Touchscreen" is attached at the end of /sys/class/input/event?/device/name
according to hidinput_allocate() in <linux source code tree>/drivers/hid/hid-input.c

There is no performance degradation because sysfs is checked only on the
first call and state is kept in an internal static variable and referred
afterward.

Signed-off-by: Caz Yokoyama <caz@caztech.com>

CazYokoyama commented:
Add touchscreen checking. See its commit message for detail.

4 ways of interacting with xcsoar
-------------------
I did not know Gesture (up / down). We may not need zoom in/out button because Gesture is much easier, no need to point a specific area.
I was surprised when I push Volume up button on a Phone. It zooms in instead of volume up when I find warning sound is too weak. I have to get out xcsoar, volume up and invoke xcsoar again. Not good.

CazYokoyama commented:
Codacy static code analysis reports

- #include <fcntl.h> - is not need because stdlib. But fail Raspbian(rpi). fcntl.h is included many other places.
- count = read(eventf, dname, sizeof(dname)); - need boundary check. No idea how to include boundary check.

I will be glad if you give me an advice.

lordfolken commented:
> Add touchscreen checking. See its commit message for detail.
> 
> ## 4 ways of interacting with xcsoar
> I did not know Gesture (up / down). We may not need zoom in/out button because Gesture is much easier, no need to point a specific area. 

I know we have a dscoverability Problem #1546 

> I was surprised when I push Volume up button on a Phone. It zooms in instead of volume up when I find warning sound is too weak. I have to get out xcsoar, volume up and invoke xcsoar again. Not good.

I agree with you, we should change the volume buttons.     Zooming without panning ability, doesnt make sense here.

lordfolken commented:
https://github.com/XCSoar/XCSoar/pull/1867 @CazYokoyama what do you think?

CazYokoyama commented:
> #1867 @CazYokoyama what do you think?

I enables menu and zoom buttons and use a lot even I know I can do it by gestures because I am Tophat user for a long time. I am glad if you keep those buttons with enabling/disabling capability.

lordfolken reviewed (commented):
## Review notes

Thanks for tackling Raspberry Pi touch detection when running without X — the sysfs fallback is a reasonable approach when `UI::event_queue->HasTouchScreen()` stays false.

### `[[gnu::pure]]` vs. cached state

`HasTouchScreen()` is still declared `[[gnu::pure]]` in `Asset.hpp`. A `pure` function must not modify global/static state, but the new implementation updates a `static` cache. That mismatch can confuse the optimizer and contradicts the attribute contract.

**Suggestion:** Drop `[[gnu::pure]]` on `HasTouchScreen()` for this configuration, or avoid mutable cache (with the usual tradeoff of repeated sysfs walks).

### Includes

The `.cpp` uses `open`, `read`, `close`, `snprintf`, and `strstr`. Elsewhere in the tree these typically get explicit includes, e.g. `<unistd.h>` for POSIX read/close, `<cstdio>` for `snprintf`, `<cstring>` for `strstr`, plus `<fcntl.h>` as you already have.

### Heuristic

Matching the substring `"TouchScreen"` in `/sys/class/input/event*/device/name` is hardware/driver-specific. Consider a one-line comment naming which Pi touch setup or kernel naming this targets so future readers know when the fallback applies or needs extending.

### Minor nits

- Extra space in `i <  MAX_EVENT_DEVICES`.
- Alignment of the `snprintf` continuation with the surrounding style (tabs/spaces) for consistency.

Happy to re-review after tweaks.

lordfolken reviewed (changes requested):
Thanks for tackling Raspberry Pi touch detection when running without X — the sysfs fallback is a reasonable approach when `UI::event_queue->HasTouchScreen()` stays false.

Below is an updated review against current `master`, with the project PR template and checkboxes.

---

## Pre-Submission Checklist

### Git & Commit History

#### Branch & Rebase
- [ ] PR is rebased on current `master`
- [ ] Use `git rebase -i` to clean up commit history before PR submission

#### Commit Format & Messages
- [ ] All commits follow the format: `<Component>: <Summary>` (no `src/` prefix)
- [ ] Use present tense in commit messages ("Fix" not "Fixed", "Add" not "Added")
- [x] Commit message body explains *why* the change was made (console without X, sysfs naming, cache)
- [ ] Commit subject uses project format (current: `[menu zoom button] add checking…`)

#### Commit Structure
- [x] Single file, one logical change
- [x] Self-contained (touch detection only)
- [x] No unrelated drive-by changes
- [ ] Each commit builds on current `master` after rebase

---

- [ ] I'm ready to merge

**Note:** CI was green on the PR branch (Apr 2025), but the branch predates `HasTouchScreenHardware()` + `CommandLine::ApplyTouchInputOverride()` on `master`. Rebase and re-run CI.

---

## What's already good

- [x] **Surgical scope** — only `Asset.cpp`; no unrelated changes
- [x] **Real problem** — `HasTouchScreen()` drives layout/zoom hit areas; false on Pi console breaks UX
- [x] **Sysfs cache** — one-time scan, then static state; no per-frame cost
- [x] **Tries event queue first** — sysfs only when `UI::event_queue->HasTouchScreen()` is false
- [x] **Heuristic matches reported hardware** — `"TouchScreen"` fits Goodix (`Goodix Capacitive TouchScreen`)
- [x] **Commit body documents intent** (hid-input naming, performance)

---

## Blockers (must fix before merge)

- [ ] **Rebase onto current `master`** — touch detection is now split:

  - `HasTouchScreenHardware()` — device detection
  - `HasTouchScreen()` — wraps with `CommandLine::ApplyTouchInputOverride()`

  Put the sysfs fallback in **`HasTouchScreenHardware()`**, not by replacing `HasTouchScreen()` wholesale, so `-touchscreen` / `-notouchscreen` keep working.

- [ ] **`open()` failure must `continue`, not `break`** — if `event0` is missing but `event2` exists, `break` stops the loop and never checks later nodes. Gaps in `/sys/class/input/eventN` are common.

- [ ] **`[[gnu::pure]]` vs static cache** — `HasTouchScreen()` is still declared `[[gnu::pure]]` in `Asset.hpp`, but this implementation mutates a `static` cache. Drop `pure` for this configuration, or avoid mutable static state.

---

## Should fix

- [ ] **Scope** — code runs for all `USE_LIBINPUT || USE_WAYLAND` builds, not only Raspberry Pi. Either guard with `#ifdef RASPBERRY_PI` or document sysfs as a general Linux console fallback.
- [ ] **Includes** — add explicit headers: `<unistd.h>` (`read`, `close`), `<cstdio>` (`snprintf`), `<cstring>` (`strstr`); keep `<fcntl.h>`.
- [ ] **Prefer `constexpr` over `#define`** for limits/paths where practical (project style).
- [ ] **Heuristic** — `strstr(..., "TouchScreen")` is case-sensitive; consider case-insensitive match or additional substrings.
- [ ] **Enumeration** — scanning `event0`…`event19` is fragile; consider `opendir("/sys/class/input")` on `event*`, or document why sequential scan is sufficient on Pi.
- [ ] **Comment** — one line naming target hardware/kernel (e.g. Goodix Pi panel) so future readers know when to extend the match.
- [ ] **`NEWS.txt`** — brief UI/embedded note if this fixes Pi zoom/menu touch sizing.

---

## Minor nits

- [ ] Extra space: `i <  MAX_EVENT_DEVICES`
- [ ] `snprintf` continuation indent (tabs/spaces) for consistency with surrounding code

---

## Design question

Default Pi builds use `USE_LIBINPUT=y`. If libinput/udev can be made to report touch reliably on console, that may be preferable long term — but sysfs fallback is still reasonable when enumeration fails or is late.

---

Happy to re-review after rebase and the blockers above.
